File: src/om/platform/MindXOM_SDK/src/cpp/devm/devm.c
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2025-2025. All rights reserved.
   OMSDK is licensed under Mulan PSL v2.
   You can use this software according to the terms and conditions of the Mulan PSL v2.
   You may obtain a copy of Mulan PSL v2 at:
            http://license.coscl.org.cn/MulanPSL2
   THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
   EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
   MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
   See the Mulan PSL v2 for more details.
 * Description: CleanCode增加devm.c描述
 */
#include <stdlib.h>
#include <dlfcn.h>
#include <securec.h>
#include "base_type.h"
#include "ens_base.h"
#include "ens_api.h"
#include "ens_log.h"
#include "file_checker.h"

#define DEVM_DRIVER_LIB_NAME "/usr/local/lib/libsystem_adapter.so"
#define INT_SIZE 4
#define ALARM_ID 65540

typedef int (*PROCESS_ALARM_EVENT_FUNC)(int alarm_num);
int (*g_pfn_fm_subscribe_alarm_event)(PROCESS_ALARM_EVENT_FUNC p_func);

static int process_alarm_event_func(int alarm_level)
{
    int ret;
    int fd;
    int (*dev_load)(void);
    int (*dev_open)(char *, int *);
    int (*set_attribute)(int, unsigned int, unsigned char *);
    ENS_LOG_INFO("alarm_level: %d", alarm_level);
    void *p_devm_driver_lib = safety_dlopen(DEVM_DRIVER_LIB_NAME, RTLD_GLOBAL | RTLD_LAZY, COMMON_TRUE, ROOT_UID);
    if (p_devm_driver_lib == NULL) {
        ENS_LOG_INFO("failed to open %s, devm driver APIs will not be available", DEVM_DRIVER_LIB_NAME);
        return -1;
    }

    dev_load = dlsym(p_devm_driver_lib, "dev_load");
    dev_open = dlsym(p_devm_driver_lib, "dev_open");
    set_attribute = dlsym(p_devm_driver_lib, "set_attribute");
    if (dev_load == NULL || dev_open == NULL || set_attribute == NULL) {
        ENS_LOG_INFO("failed to load set alarm level function.");
        dlclose(p_devm_driver_lib);
        return -1;
    }

    ret = dev_load();
    if (ret != 0) {
        ENS_LOG_INFO("failed to call dev load function.");
        dlclose(p_devm_driver_lib);
        return -1;
    }

    ret = dev_open("system0", &fd);
    if (ret != 0) {
        ENS_LOG_INFO("failed to call dev open function.");
        dlclose(p_devm_driver_lib);
        return -1;
    }

    unsigned int buffSize = INT_SIZE * 3;
    unsigned char *buff = (unsigned char *)malloc(buffSize);
    if (buff == NULL) {
        ENS_LOG_INFO("call malloc fail.");
        dlclose(p_devm_driver_lib);
        return -1;
    }
    *(int *)(buff) = ALARM_ID;
    *(int *)(buff + INT_SIZE) = INT_SIZE;
    *(int *)(buff + INT_SIZE * 2) = alarm_level;
    ret = set_attribute(fd, buffSize, buff);
    if (ret != 0) {
        ENS_LOG_INFO("failed to call set alarm level function.");
        dlclose(p_devm_driver_lib);
        free(buff);
        return -1;
    }

    free(buff);
    return ret;
}

int devm_load(void)
{
    int ret = ens_intf_import("subscribe_alarm_event", (void **)&g_pfn_fm_subscribe_alarm_event);
    ENS_LOG_INFO("devm load ret:%d", ret);
    return ENS_OK;
}

int devm_unload(void)
{
    ENS_LOG_WARN("devm unload,but do nothing now!");
    return ENS_OK;
}

int devm_start(void)
{
    if (g_pfn_fm_subscribe_alarm_event == NULL) {
        ENS_LOG_ERR("g_pfn_fm_subscribe_alarm_event null");
        return ENS_ERR;
    }

    int ret = g_pfn_fm_subscribe_alarm_event(process_alarm_event_func);
    ENS_LOG_INFO("subscribe_alarm_event ret:%d", ret);
    return ENS_OK;
}

int devm_stop(void)
{
    ENS_LOG_WARN("devm stop,but do nothing now!");
    return ENS_OK;
}

File: src/om/platform/MindXOM_SDK/src/app/sys/lib/Linux/systems/devm.py
# -*- coding: utf-8 -*-
# Copyright (c) Huawei Technologies Co., Ltd. 2025-2025. All rights reserved.
# OMSDK is licensed under Mulan PSL v2.
# You can use this software according to the terms and conditions of the Mulan PSL v2.
# You may obtain a copy of Mulan PSL v2 at:
#          http://license.coscl.org.cn/MulanPSL2
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
# EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
# MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
# See the Mulan PSL v2 for more details.

import threading

from common.log.logger import run_log
from common.utils.common_check import CommonCheck
from common.common_methods import CommonMethods
from devm.device_mgr import DEVM
from devm.exception import DeviceManagerError


class Module:

    def __init__(self):
        """
        功能描述：初始化函数
        参数：
        返回值：无
        异常描述：NA
        """
        # 初始化模组列表
        self.items = []
        self.devices = []
        self.ModuleInfo = {}
        self.get_module_list()

    def get_all_info(self, module_id=None) -> list:
        """
        功能描述：获取模组资源列表以及指定模组信息
        参数：module_id 模组名 为None是获取模组列表
        返回值：list
        异常描述：NA
        """
        self.get_module_list()
        if module_id is None:
            return [0, ]

        if module_id not in self.items:
            run_log.error("module %s not exist", module_id)
            return [CommonMethods.NOT_EXIST, '{} does not exist'.format(module_id)]

        self.get_module_info(module_id)
        run_log.info("get module %s info success", module_id)
        return [CommonMethods.OK, "get module info success"]

    def get_module_list(self):
        try:
            # 目前获取的模组列表仅包含拓展模组
            self.items = DEVM.get_module_list_by_category('addition')
        except DeviceManagerError:
            run_log.error("failed to get module list.")

    def get_module_info(self, module_id):
        try:
            self.devices = DEVM.get_device_list_by_module(module_id)
        except DeviceManagerError:
            run_log.error("failed to get device list, module name is %s", module_id)
            return

        try:
            self.ModuleInfo = DEVM.get_module_info(module_id)
        except DeviceManagerError:
            run_log.error("failed to get module info, module name is %s", module_id)


class Device:
    DEVICE_LOCK = threading.Lock()

    def __init__(self):
        """
        功能描述：初始化函数
        参数：
        返回值：无
        异常描述：NA
        """
        self.Attributes = {}

    @staticmethod
    def set_device_info(device_id, attributes: dict):
        for attribute_name, value in attributes.items():
            try:
                DEVM.get_device(device_id).set_attribute(attribute_name, value)
            except DeviceManagerError:
                run_log.error("failed to set device %s attribute %s" % device_id, attribute_name)
                return

    @staticmethod
    def _check_if_device_belongs_to_module(device_id, module_id):
        return device_id in DEVM.get_device_list_by_module(module_id)

    def get_all_info(self, module_id=None, device_id=None) -> list:
        """
        功能描述：获取模组资源列表以及指定模组信息
        参数：module_id 模组名 device_id 设备名
        返回值：list
        异常描述：NA
        """
        module_obj = Module()
        if module_id not in module_obj.items:
            run_log.error("module %s not exist", module_id)
            return [CommonMethods.NOT_EXIST, '{} does not exist'.format(module_id)]

        if not self._check_if_device_belongs_to_module(device_id, module_id):
            run_log.error("module %s has no device %s", module_id, device_id)
            return [CommonMethods.NOT_EXIST, f"module '{module_id}' has no device '{device_id}'"]

        self.get_device_info(device_id)
        run_log.info("get device %s attributes success", device_id)
        return [CommonMethods.OK, "get device info success"]

    def patch_request(self, request_data, module_id=None, device_id=None):
        """
        功能描述：patch请求的处理参数
        参数：request_data  请求参数
             module_id为模组名称，device_id为设备名称
        返回值：是否成功的列表信息
        异常描述：
        """
        module_obj = Module()
        if module_id not in module_obj.items:
            run_log.error("module %s not exist", module_id)
            return [CommonMethods.NOT_EXIST, '{} does not exist'.format(module_id)]

        if not self._check_if_device_belongs_to_module(device_id, module_id):
            run_log.error("module %s has no device %s", module_id, device_id)
            return [CommonMethods.NOT_EXIST, f"module '{module_id}' has no device '{device_id}'"]

        if Device.DEVICE_LOCK.locked():
            run_log.warning("Device modify is busy")
            return [CommonMethods.ERROR, "Device modify is busy"]

        with Device.DEVICE_LOCK:
            try:
                operator_check = CommonCheck.check_operator(request_data.get("_User"), request_data.get("_Xip"))
                if not operator_check:
                    run_log.error("The operator is illegal, %s", operator_check.error)
                    return [CommonMethods.ERROR, "The operator is illegal."]

                self.set_device_info(device_id, request_data.get("Attributes", {}))
                run_log.info("set device %s info success, module name : %s", device_id, module_id)
                self.get_device_info(device_id)
                run_log.info("get device %s info success", device_id)
                return [CommonMethods.OK, "Device modify success"]
            except Exception as err:
                run_log.error("Device modify failed, reason is %s", (str(err)))
                return [CommonMethods.ERROR, "Device modify failed, reason is {}".format(str(err))]

    def get_device_info(self, device_id):
        try:
            self.Attributes = DEVM.get_device(device_id).get_all_attributes()
        except DeviceManagerError:
            run_log.error("failed to get device %s attributes." % device_id)
